code/espurna/sensors/EZOPHSensor.h

```diff
@@ -0,0 +1,210 @@
+        }
+
+        // ---------------------------------------------------------------------
+
+        unsigned char getRX() {
+            return _pin_rx;
+        }
+
+        unsigned char getTX() {
+            return _pin_tx;
+        }
+
+        // ---------------------------------------------------------------------
+        // Sensor API
+        // ---------------------------------------------------------------------
+
+        // Initialization method, must be idempotent
+        void begin() {
+            if (!_dirty) return;
+
+            if (_serial) delete _serial;
+
+            _serial = new SoftwareSerial(_pin_rx, _pin_tx);
+            _serial->enableIntTx(false);
+            _serial->begin(9600);
+
+            _ready = true;
+            _dirty = false;
+        }
+
+        // Descriptive name of the sensor
+        String description() {
+            char buffer[28];
+            snprintf(buffer, sizeof(buffer), "EZOPH @ SwSerial(%u,%u)", _pin_rx, _pin_tx);
+            return String(buffer);
+        }
+
+        // Descriptive name of the slot # index
+        String slot(unsigned char index) {
+            return description();
+        };
+
+        // Address of the sensor (it could be the GPIO or I2C address)
+        String address(unsigned char index) {
+            char buffer[6];
+            snprintf(buffer, sizeof(buffer), "%u:%u", _pin_rx, _pin_tx);
+            return String(buffer);
+        }
+
+        // Type for slot # index
+        unsigned char type(unsigned char index) {
+            if (index == 0) return MAGNITUDE_PH;
+            return MAGNITUDE_NONE;
+        }
+
+        void tick() {
+            _setup();
+            _read();
+        }
+
+        // Current value for slot # index
+        double value(unsigned char index) {
+            if (index == 0) return _ph;
+            return 0;
+        }
+
+
+
+    protected:
+
+        // ---------------------------------------------------------------------
+        // Protected
+        // ---------------------------------------------------------------------
+
+        void _setup() {
+          if (_sync_responded == true) {
+            return;
+          }
+
+          _error = SENSOR_ERROR_WARM_UP;
+
+          String sync_serial = "";
+          sync_serial.reserve(30);
+
+          if (_sync_requested == false) {
+              _serial->write(67); // C
+              _serial->write(44); // ,
+              _serial->write(63); // ?
+              _serial->write(13); // \r
+
+              _sync_requested = true;
+          }
+
+          while ((_serial->available() > 0)) {
+              char sync_char = (char)_serial->read();
+              sync_serial += sync_char;
+
+              if (sync_char == '\r') {
+                break;
+              }
+          }
+
+          if (sync_serial.startsWith("?C,")) {
+              _sync_interval = sync_serial.substring(sync_serial.indexOf(",") + 1).toInt() * 1000;
```

Comment:
indexof("?C,")?

Reply:
I'm already making sure the string starts with `?C,` and that is the data format they use so I think it's good as it is. Alternative would be `indexOf("?C,") + 3` which wouldn't look as good.